quint/src/ToIrListener.ts

```diff
@@ -591,19 +600,66 @@ export class ToIrListener implements QuintListener {
     this.pushApplication(ctx, 'List', args)
   }
 
-  // record constructor, e.g., { name: "igor", year: 2021 }
-  exitRecord(ctx: p.RecordContext) {
-    const names = ctx.IDENTIFIER().map(n => n.text)
-    const elems: QuintEx[] = popMany(this.exprStack, ctx.expr().length)
-
-    const namesAndValues = zipWith(names, elems, (name, elem) => {
+  // field: expr, or ...record
+  exitRecElem(ctx: p.RecElemContext) {
+    const expr = this.exprStack.pop()!
+    if (ctx.IDENTIFIER()) {
+      // field: expr
       const id = this.idGen.nextId()
       this.sourceMap.set(id, this.loc(ctx))
+      const nameEx: QuintEx = {
+        id,
+        kind: 'str',
+        value: ctx.IDENTIFIER()?.text!,
+      }
+      // the 'Tup' id is not needed, as the tuple is unwrapped
+      this.exprStack.push({
+        id: 0n,
+        kind: 'app',
+        opcode: 'Tup',
+        args: [nameEx, expr],
+      })
+    } else {
+      // ...expr
+      // the 'Tup' id is not needed, as the tuple is unwrapped
+      this.exprStack.push({
+        id: 0n,
+        kind: 'app',
+        opcode: 'Tup',
+        args: [expr],
+      })
+    }
+  }
 
-      const nameExpression: QuintEx = { id, kind: 'str', value: name }
-      return [nameExpression, elem]
-    })
-    this.pushApplication(ctx, 'Rec', namesAndValues.flat())
+  // record constructor, e.g., { name: "igor", year: 2021 }
+  exitRecord(ctx: p.RecordContext) {
+    const elems = popMany(this.exprStack, ctx.recElem().length)
+    const spreads = elems.filter(e => e.kind === 'app' && e.args.length === 1)
+    const pairs = elems.map(e => (e.kind === 'app' && e.args.length === 2 ? e.args : [])).filter(es => es.length > 0)
+
+    if (spreads.length === 0) {
+      // { field1: value1, field2: value2 }
+      this.pushApplication(ctx, 'Rec', pairs.flat())
+    } else if (spreads.length > 1) {
+      // error
+      const msg = 'QNT012: ... may be used once in { ...record, <fields> }'
+      this.pushError(ctx, msg)
```

Comment:
If we were using the `{r with f1: n}` syntax this would be made impossible by the grammar.

Reply:
This is true. However, it makes no difference for the end user.